feat: Implement `IntoEngineData` for `DomainMetadata`

lbhm opened the pull request:
## What changes are proposed in this pull request?

This PR implements the `IntoEngineData` trait for `DomainMetadata` actions, which is a prerequisite for implementing the row tracking write path.

This PR also adds `Scalar::try_from` implementations for `HashMap` and `Vec` and consequently simplifies existing implementations of `IntoEngineData`.

## How was this change tested?

New UT

scovich commented on kernel/src/actions/mod.rs:
```diff
@@ -740,6 +740,60 @@ impl Add {
+        .map(Scalar::Map)?;
+
+        let tags = MapData::try_new(
+            MapType::new(DataType::STRING, DataType::STRING, false),
```
If null tag and missing tag are semantically different, then we will need to expand the row visitor infrastructure to handle that situation and update the type of the field to `HashMap<String, Option<String>>`. Can we file that as a follow-up item?

lbhm commented:
@scovich I addressed your comments and added more `Scalar::from` implementations to derive as many `IntoEngineData` implementations as feasible.

IMO, this PR adds value independent of our overall decision on row tracking, so I would vote for making an approval decision based on its own scope.

lbhm commented on kernel/src/expressions/scalars.rs:
```diff
@@ -545,6 +546,70 @@ impl From<&[u8]> for Scalar {
+    }
+}
+
+impl<K, V> From<HashMap<K, V>> for Scalar
+where
+    K: Into<Scalar> + ToDataType,
+    V: Into<Scalar> + ToDataType,
+{
+    fn from(map: HashMap<K, V>) -> Self {
+        let key_type = K::to_data_type();
+        let value_type = V::to_data_type();
+        let map_type = MapType::new(key_type, value_type, false);
+
+        let map_data =
+            MapData::try_new(map_type, map).expect("Failed to create MapData from HashMap");
+
+        Self::Map(map_data)
+    }
+}
```
After moving to `TryFrom` as suggested by @zachschuermann, the compiler complains about a conflicting implementation when I refactor this to `IntoIterator`, so I'm not sure if we can do that. IIUC, the issue is that we want to implement `TryFrom` but not `From`.

```rust
impl<T, U> TryFrom<U> for T
  where U: Into<T>;
```

OussamaSaoudi commented on kernel/src/actions/mod.rs:
```diff
@@ -289,42 +289,20 @@ impl IntoEngineData for Metadata {
-            partition_columns,
-            created_time,
-            configuration,
+            self.id.into(),
+            self.name.into(),
+            self.description.into(),
+            self.format.provider.into(),
+            self.format.options.try_into()?,
+            self.schema_string.into(),
+            self.partition_columns.try_into()?,
+            self.created_time.into(),
+            self.configuration.try_into()?,
```
this looks really clean. Nice!

OussamaSaoudi commented on kernel/src/expressions/scalars.rs:
```diff
@@ -999,4 +1068,187 @@ mod tests {
+        assert_eq!(scalar.data_type(), DataType::from(expected_map_type));
+
+        // Extract the MapData and verify contents
+        if let Scalar::Map(map_data) = scalar {
+            let pairs = map_data.pairs();
+            assert_eq!(pairs.len(), 2);
+
+            // Check that both expected pairs are present
+            let has_key1 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
+            });
+            let has_key2 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Integer(100))
+            });
+
+            assert!(has_key1, "Missing key1 -> 42 pair");
+            assert!(has_key2, "Missing key2 -> 100 pair");
+        } else {
+            panic!("Expected Map scalar");
+        }
```
nit: We can return early and remove the nesting
```suggestion
        let Scalar::Map(map_data) = scalar else {
            panic!("Expected Map scalar");
        }
        let pairs = map_data.pairs();
        assert_eq!(pairs.len(), 2);

        // Check that both expected pairs are present
        let has_key1 = pairs.iter().any(|(k, v)| {
            matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
        });
        let has_key2 = pairs.iter().any(|(k, v)| {
            matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Integer(100))
        });

        assert!(has_key1, "Missing key1 -> 42 pair");
        assert!(has_key2, "Missing key2 -> 100 pair");
```

OussamaSaoudi commented on kernel/src/expressions/scalars.rs:
```diff
@@ -999,4 +1068,187 @@ mod tests {
+        assert_eq!(scalar.data_type(), DataType::from(expected_map_type));
+
+        // Extract the MapData and verify contents
+        if let Scalar::Map(map_data) = scalar {
+            let pairs = map_data.pairs();
+            assert_eq!(pairs.len(), 3);
+
+            // Check that all expected pairs are present
+            let has_key1 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
+            });
+            let has_key2 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Null(_))
+            });
+            let has_key3 = pairs.iter().any(|(k, v)| {
+                matches!(k, Scalar::String(s) if s == "key3") && matches!(v, Scalar::Integer(100))
+            });
+
+            assert!(has_key1, "Missing key1 -> 42 pair");
+            assert!(has_key2, "Missing key2 -> null pair");
+            assert!(has_key3, "Missing key3 -> 100 pair");
+
+            // Verify the map type allows null values
+            assert!(map_data.map_type().value_contains_null());
+        } else {
+            panic!("Expected Map scalar");
+        }
```
similar here, return early with a `let Scalar::Map(map_data) = scalar else {panic}`
```suggestion
       let Scalar::Map(map_data) = scalar else {
             panic!("Expected Map scalar");
       }
      let pairs = map_data.pairs();
      assert_eq!(pairs.len(), 3);

      // Check that all expected pairs are present
      let has_key1 = pairs.iter().any(|(k, v)| {
          matches!(k, Scalar::String(s) if s == "key1") && matches!(v, Scalar::Integer(42))
      });
      let has_key2 = pairs.iter().any(|(k, v)| {
          matches!(k, Scalar::String(s) if s == "key2") && matches!(v, Scalar::Null(_))
      });
      let has_key3 = pairs.iter().any(|(k, v)| {
          matches!(k, Scalar::String(s) if s == "key3") && matches!(v, Scalar::Integer(100))
      });

      assert!(has_key1, "Missing key1 -> 42 pair");
      assert!(has_key2, "Missing key2 -> null pair");
      assert!(has_key3, "Missing key3 -> 100 pair");

      // Verify the map type allows null values
      assert!(map_data.map_type().value_contains_null());
```

nicklan reviewed (approved):
this lgtm thanks! agree with Ryan's nits so let's just fix those and then we can merge

lbhm commented on kernel/src/actions/mod.rs:
```diff
@@ -982,6 +938,25 @@ mod tests {
+    ) -> MapBuilder<StringBuilder, StringBuilder> {
+        MapBuilder::new(
+            Some(MapFieldNames {
+                entry: "key_value".to_string(),
```
TBH, I just copied this code snippet and factored it out into a dedicated function. This pattern is used consistently throughout the code base. Just replacing `Some(MapFieldNames(...))` with `None` results in errors, so I would argue against addressing this in the scope of this PR.
